dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+      * Causes too many overloads if we accept `Task` and `void`
+  * Optional `Name` positional property can be given. Default is the unqualified method name with "Async" trimmed off if
+    present.
+* Workflow queries are public methods with the non-inheritable `[WorkflowQuery]` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method cannot return a `Task`, but can accept arguments
+  * Optional `Name` positional property can be given. Default is the unqualified method name.
+  * Would love to put this on properties, but they aren't referenceable as delegates
+* Users are encouraged to set a `static` `readonly` `Ref` property for use by clients
+  * `Refs.Create<T>()` is how it is obtained
+    * `Refs` is in `Temporalio` not `Temporalio.Workflows`
+      * TODO(cretz): This acceptable?
+    * TODO(cretz): Should we do validation at this time on the attributes?
+      * Problem is instance activities will want to use `Refs` too.
+      * If I put `Refs` in `Temporalio.Workflows` then at least I can know that all uses are for workflows
+        * But then it'd clash with a `Temporalio.Activities.Refs` or do we want something like either
+          `Refs.ForWorkflow<T>` or `WorkflowRefs.Create<T>`?
+  * This is backed by a dynamic proxy
+    * Since the constructor may be `[WorkflowInit]`, we use
+      `System.Runtime.Serialization.FormatterServices.GetUninitializedObject`, ref
+      [this issue](https://github.com/castleproject/Core/issues/588)
+* Notice all the cool uses of `System.Threading.Tasks.Dataflow`
+
+## .NET Workflow Worker
+
+Plan:
+
+* Create `WorkflowInstance` that is a custom .NET `TaskScheduler` that executes deterministically, until all are
+  idle, and properly executes wait-condition and such
+  * The scheduler will be the current scheduler and no workflow work is allowed on any other scheduler
+  * Each task run will be synchronous on a single thread so deadlock detection can work
+  * All task scheduling should then work, e.g. `Task.Delay`, Dataflow stuff, etc
+* Follow Python and others' example of task ordering/distribution/etc
+* Not creating a `WorkflowRunner` abstraction like in Python because we have no reasonable expectation for a sandbox in
+  the future
+  * This abstraction can be created later, but since .NET does not have AppDomain anymore, there is no way to build a
+    good sandbox, so we will rely on custom analyzers instead
+
+Additions to `TemporalWorkerOptions`:
+
+```csharp
+namespace Temporalio.Worker;
+
+public class TemporalWorkerOptions : ICloneable
+{
+    // ... phase 1 pieces omitted ...
+
+    public IList<Type> Workflows { get; set; } = new List<Type>();
+
+    public TemporalWorkerOptions AddWorkflow(Type type) { /*... */ }
+
+    public int MaxCachedWorkflows { get; set; } = 10000;
+
+    public int MaxConcurrentWorkflowTasks { get; set; } = 100;
+
+    public double NonStickyToStickyPollRatio { get; set; } = 0.2;
```

Comment:
I'd avoid exposing this for now since we might remove this option in the future from Core.

Reply:
Fixed